src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,672 @@
+              ++(*parsed_tz_ptr);
+            *parsed_tz_length = end_ptr - *parsed_tz_ptr;
+            break;
+          }
+          if (i == 6 && '.' != b) { i += 1; }
+        } else {
+          if (i < segments_len && (':' == b || ' ' == b)) {
+            if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+          } else {
+            return ParseResult::INVALID;
+          }
+        }
+      } else {
+        if (6 == i) { digits_milli += 1; }
+        // We will truncate the nanosecond part if there are more than 6 digits,
+        // which results in loss of precision
+        if (6 != i || current_segment_digits < 6) {
+          current_segment_value = current_segment_value * 10 + parsed_value;
+        }
+        current_segment_digits += 1;
+      }
+      j += 1;
+    }
+
+    if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+    segments[i] = current_segment_value;
+
+    while (digits_milli < 6) {
+      segments[6] *= 10;
+      digits_milli += 1;
+    }
+
+    segments[0] *= year_sign.value_or(1);
+    // above is ported from Spark.
+
+    // copy segments to equivalent kernel timestamp_components
+    // Note: In order to keep above code is equivalent to Spark implementation,
+    //       did not use `timestamp_components` directly to save values.
+    ts_comp->year         = segments[0];
+    ts_comp->month        = static_cast<int8_t>(segments[1]);
+    ts_comp->day          = static_cast<int8_t>(segments[2]);
+    ts_comp->hour         = static_cast<int8_t>(segments[3]);
+    ts_comp->minute       = static_cast<int8_t>(segments[4]);
+    ts_comp->second       = static_cast<int8_t>(segments[5]);
+    ts_comp->microseconds = segments[6];
+
+    return ParseResult::OK;
+  }
+};
+
+/**
+ * The common entrance of string_to_timestamp, two paths call this function:
+ * - `string_to_timestamp_with_tz` : with time zone
+ * - `string_to_timestamp_without_tz` : without time zone
+ * The parameters transitions, tz_indices and default_tz_index are only for handling
+ * inputs with timezone.
+ * It's called from `string_to_timestamp_without_tz` if transitions and tz_indices
+ * are nullptr, otherwise called from `string_to_timestamp_with_tz`.
+ *
+ */
+std::unique_ptr<cudf::column> to_timestamp(cudf::strings_column_view const& input,
+                                           bool ansi_mode,
+                                           bool allow_tz_in_date_str            = true,
+                                           size_type default_tz_index           = 1000000000,
+                                           cudf::column_view const* transitions = nullptr,
+                                           cudf::column_view const* tz_indices  = nullptr)
+{
+  auto const stream = cudf::get_default_stream();
```

Comment:
The common approach is to have a public API with no stream or mr and then a detail API that has default arguments for the values.